### PR TITLE
Add public management ctor generation

### PR DESCRIPTION
**Pull request by pakrym**
Also changes how standalone mode works - instead of passing all parameters via command line the hosted run serializes configuration into a file that standalone process later loads. This way we can use README.md to set options (title, azure-arm) and standalone invocation would see them.

### CODE REVIEW

File: src/AutoRest.CSharp.V3/Generation/Writers/ClientWriter.cs [hunk 1]
```diff
@@ -58,6 +67,56 @@ public void WriteClient(CodeWriter writer, Client client)
             }
         }
 
+        private void WriteManagementClientCtors(CodeWriter writer, Client client, Configuration configuration)
+        {
```
Comment (pakrym) [inline src/AutoRest.CSharp.V3/Generation/Writers/ClientWriter.cs:71]: This is pretty hardcoded but should work for the first iteration, will improve as we gather more feedback.

File: samples/Azure.Storage.Management/readme.md [hunk 2]
```diff
@@ -4,7 +4,8 @@
 > see https://aka.ms/autorest
 
 ``` yaml
-title: Azure.Storage.Management
+title: Storage
```
Comment (MiYanni) [inline samples/Azure.Storage.Management/readme.md:7]: So... are we just using folder names to separate things between management and data plane?

Comment (MiYanni) [inline samples/Azure.Storage.Management/readme.md:7]: So... are we just using folder names to separate things between management and data plane?

File: src/AutoRest.CSharp.V3/AutoRest/Communication/IPluginCommunication.cs [hunk 3]
```diff
@@ -3,15 +3,15 @@
...
+using AutoRest.CSharp.V3.AutoRest.Plugins;
 
 namespace AutoRest.CSharp.V3.AutoRest.Communication
 {
     internal interface IPluginCommunication
     {
         string PluginName { get; }
-        Task<string> ReadFile(string filename);
-        Task<T> GetValue<T>(string key);
-        Task<string[]> ListInputs(string? artifactType = null);
```
Comment (MiYanni) [inline src/AutoRest.CSharp.V3/AutoRest/Communication/IPluginCommunication.cs:14]: So, these were AutoRest communication concepts (meaning, the JsonRPC API for AutoRest). So, changing them to non-AutoRest related methods would make us not have plugin communication (multiple plugins) in the future if we wanted/needed it.

The interface here was only so other plugin communcation methods (such as our Host communication) could communicate in the same way any other AutoRest plugin would.

File: src/AutoRest.CSharp.V3/AutoRest/Communication/IPluginCommunication.cs [hunk 4]
```diff
@@ -3,15 +3,15 @@
...
+        Configuration Configuration { get; }
+        Task<string> GetCodeModel();
```
Comment (MiYanni) [inline src/AutoRest.CSharp.V3/AutoRest/Communication/IPluginCommunication.cs:14]: This is very specific to the plugin pipeline, and not related to plugin communication.

File: src/AutoRest.CSharp.V3/AutoRest/Plugins/ClientOptionsWriter.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+using System.Linq;
+using AutoRest.CSharp.V3.Generation.Writers;
+using AutoRest.CSharp.V3.Output.Models.Types;
+using Azure.Core;
+
+namespace AutoRest.CSharp.V3.AutoRest.Plugins
+{
+    internal class ClientOptionsWriter
+    {
+        public static void WriteClientOptions(CodeWriter writer, BuildContext context)
```
Comment (MiYanni) [inline src/AutoRest.CSharp.V3/AutoRest/Plugins/ClientOptionsWriter.cs:14]: What is this for?

File: src/AutoRest.CSharp.V3/AutoRest/Communication/IPluginCommunication.cs (hunk 3 above)
Comment (pakrym) [reply src/AutoRest.CSharp.V3/AutoRest/Communication/IPluginCommunication.cs:14]: Yeah, you are right, I'll redo this change without messing up this abstraciton.

File: src/AutoRest.CSharp.V3/AutoRest/Plugins/ClientOptionsWriter.cs (hunk 5 above)
Comment (pakrym) [reply src/AutoRest.CSharp.V3/AutoRest/Plugins/ClientOptionsWriter.cs:14]: Generates ***ClientOptions classes that are required for public ctors. https://github.com/Azure/autorest.csharp/pull/644/files#diff-1facde7a0e058c6acf8cc086f0635dc8

Comment (MiYanni) [reply src/AutoRest.CSharp.V3/AutoRest/Plugins/ClientOptionsWriter.cs]: Got it. All clients we generate use the same **ClientOptions, so it makes sense to generate it with the name of the library, and not the name of the client itself. For example, it isn't UsagesClientOptions, but instead StorageManagementClientOptions for any client in StorageManagement.

File: eng/Generate.ps1 [hunk 6]
```diff
@@ -141,11 +140,11 @@ if ($updateLaunchSettings)
         $definition = $swaggerDefinitions[$key];
         $outputPath = (Join-Path $definition.output $key).Replace($repoRoot, '$(SolutionDir)')
         $codeModel = Join-Path $outputPath 'CodeModel.yaml'
-        $namespace = $definition.title.Replace('-', '_')
+        $namespace = $definition.projectName.Replace('-', '_')
 
         $settings.profiles[$key] = [ordered]@{
             'commandName'='Project';
-            'commandLineArgs'="--standalone --input-codemodel=$codeModel --plugin=csharpgen --output-folder=$outputPath --namespace=$namespace --shared-source-folder=$sharedSourceNormalized --save-code-model=true"
+            'commandLineArgs'="--standalone $outputPath"
```
Comment (MiYanni) [inline eng/Generate.ps1:147]: Looking at this change here finally made me realize why this PR seems odd. So, the command line argument for AutoRest are simply AutoRest configuration values. Meaning, `standalone`, `input-codemodel`, `plugin`, `output-folder`, `namespace`, `shared-source-folder`, and `save-code-model` are all values that can be put into a `readme.md`.

This PR invents another configuration file, `Configuration.json`, that serves the same purpose of the configuration in an AutoRest `readme.md`. All the values in this new `Configuration.json` are standard AutoRest values other than `SharedSourceFolder`. But, you can simply make a value `shared-source-folder` in the `readme.md` and read it in our plugin.

So, while I think it is a good idea for each sample to have it's own configuration, we should just use `readme.md` for that purpose. That's the point of AutoRest supporting configuration in this way.

File: src/AutoRest.CSharp.V3/AutoRest/Plugins/Configuration.cs [hunk 7]
```diff
@@ -2,30 +2,32 @@
...
         {
-            OutputFolder = new Uri(GetRequiredOption(autoRest, "output-folder")).LocalPath;
-            SharedSourceFolder = new Uri(GetRequiredOption(autoRest, "shared-source-folder")).LocalPath;
-            Namespace = autoRest.GetValue<string?>("namespace").GetAwaiter().GetResult() ?? "Sample";
-            Title = autoRest.GetValue<string?>("title").GetAwaiter().GetResult() ?? "Sample";
-            SaveCodeModel = autoRest.GetValue<bool?>("save-code-model").GetAwaiter().GetResult() ?? false;
+            OutputFolder = outputFolder;
+            Namespace = ns;
+            var namespaceParts = ns.Split('.');
+            LibraryName = name ?? namespaceParts.Last();
```
[the remark this answers is not in the document]
Comment (pakrym) [reply src/AutoRest.CSharp.V3/AutoRest/Plugins/Configuration.cs:20]: Chatted offline, the namespace would have to be picked manually at the time of the project creation and is the better source of a good name.

File: src/AutoRest.CSharp.V3/AutoRest/Plugins/CSharpProj.cs [hunk 8]
```diff
@@ -28,9 +28,10 @@ internal class CSharpProj : IPlugin
 </Project>
 
 ";
-        public async Task<bool> Execute(IPluginCommunication autoRest, CodeModel codeModel, Configuration configuration)
+        public async Task<bool> Execute(IPluginCommunication autoRest)
         {
-                await autoRest.WriteFile($"{configuration.Title}.csproj", _csProjContent, "source-file-csharp");
+            var ns = await autoRest.GetValue<string>("namespace");
+            await autoRest.WriteFile($"{ns}.csproj", _csProjContent, "source-file-csharp");
```
[the remark this answers is not in the document]
Comment (pakrym) [reply src/AutoRest.CSharp.V3/AutoRest/Plugins/CSharpProj.cs:34]: Chatted offline. For track 2 projects the root namespace, project name and package name are all need to be the same so namespace becomes a pretty good indication of what the project name should be.

**Pull request merged**